runtime/runenv_events.go

```diff
@@ -95,54 +200,56 @@ func (re *RunEnv) RecordMessage(msg string, a ...interface{}) {
 	if len(a) > 0 {
 		msg = fmt.Sprintf(msg, a...)
 	}
-	evt := Event{
-		Type:    EventTypeMessage,
+	e := &Event{MessageEvent: &MessageEvent{
 		Message: msg,
-	}
-	re.logger.Info("", zap.Object("event", evt))
+	}}
+	re.logger.Info("", zap.Object("event", e))
 }
 
 func (re *RunEnv) RecordStart() {
-	evt := Event{
-		Type:   EventTypeStart,
+	e := &Event{StartEvent: &StartEvent{
 		Runenv: &re.RunParams,
-	}
+	}}
 
-	re.logger.Info("", zap.Object("event", evt))
-	re.metrics.recordEvent(&evt)
+	re.logger.Info("", zap.Object("event", e))
+	re.metrics.recordEvent(e)
 }
 
 // RecordSuccess records that the calling instance succeeded.
 func (re *RunEnv) RecordSuccess() {
-	evt := Event{
-		Type:    EventTypeFinish,
-		Outcome: EventOutcomeOK,
+	e := &Event{SuccessEvent: &SuccessEvent{TestGroupID: re.RunParams.TestGroupID}}
+	re.logger.Info("", zap.Object("event", e))
+	re.metrics.recordEvent(e)
+
+	if re.signalEventer != nil {
+		_ = re.signalEventer.SignalEvent(context.Background(), e)
 	}
-	re.logger.Info("", zap.Object("event", evt))
-	re.metrics.recordEvent(&evt)
 }
 
 // RecordFailure records that the calling instance failed with the supplied
 // error.
 func (re *RunEnv) RecordFailure(err error) {
-	evt := Event{
-		Type:    EventTypeFinish,
-		Outcome: EventOutcomeFailed,
-		Error:   err.Error(),
+	e := &Event{FailureEvent: &FailureEvent{TestGroupID: re.RunParams.TestGroupID, Error: err.Error()}}
+	re.logger.Info("", zap.Object("event", e))
+	re.metrics.recordEvent(e)
+
+	if re.signalEventer != nil {
+		_ = re.signalEventer.SignalEvent(context.Background(), e)
 	}
-	re.logger.Info("", zap.Object("event", evt))
-	re.metrics.recordEvent(&evt)
 }
 
 // RecordCrash records that the calling instance crashed/panicked with the
 // supplied error.
 func (re *RunEnv) RecordCrash(err interface{}) {
-	evt := Event{
-		Type:       EventTypeFinish,
-		Outcome:    EventOutcomeCrashed,
-		Error:      fmt.Sprintf("%s", err),
-		Stacktrace: string(debug.Stack()),
+	e := &Event{CrashEvent: &CrashEvent{
+		TestGroupID: re.RunParams.TestGroupID,
+		Error:       fmt.Sprintf("%s", err),
+		Stacktrace:  string(debug.Stack()),
+	}}
+	re.logger.Error("", zap.Object("event", e))
+	re.metrics.recordEvent(e)
+
+	if re.signalEventer != nil {
```

Comment:
nit: A cleaner way would be to initialize to a default `NilSignalEventer` that noops; then you can remove the nil checks everywhere.